vpat 16: context menu as a drag-drop alternative to move collections

abaevbog opened the pull request:
- added context menu to move a collection into another collection. Since one can drag a collection into another library, the top-level entries of the contextmenu are libraries and not collections as with itemTree (shouldn't the same be true for itemTree in fact? You can drag an item to a collection in another group but I don't think it's possible to do it with the context menu).
- refactoring of collectionTree to pull out `copyCollections()` and `copyItems()` from out of the scope of `CollectionTree.onDrop`. If a collection is moved into another library, the functionality of copying over a collection needed to be exposed to ZoteroPane without the actual drag-drop happening. `copyCollection` and `copyItems` are unchanged except for the added parameters. We could avoid the refactoring by maybe faking a `drop` event from `ZoteroPane.moveToCollection` but I thought this would be cleaner and potentially useful for in the future.


<img width="1312" alt="Screenshot 2024-07-22 at 11 19 14 PM" src="https://github.com/user-attachments/assets/af16ddad-30dd-4080-aaf1-224df694a1be">

Addresses part of #3970

abaevbog commented:
> Maybe “Change Parent”? We’d want to title-case it like other context menu items in any case.

That is a good point! I'll have the case fixed. One question on "Change Parent": if you move the collection to a different library, it doesn't technically "change" the parent but rather creates a copy in another group. Would "Change Parent" not be a bit confusing because it implies that the collection should be gone from the old library? "Move to another collection" is actually no better (maybe even more misleading) in this sense. Maybe I'm really overthinking this though

abaevbog commented:
> This could get complicated, but I guess we could put all the other libraries under a submenu labeled “Other Libraries (As Copy)”. Maybe that’s weirder.

Thanks to `Zotero.Utilities.Internal.createMenuForTarget`, it seems be be quite straightforward actually! I think this looks nicer and more informative

<img width="1148" alt="Screenshot 2024-07-26 at 5 06 17 PM" src="https://github.com/user-attachments/assets/3cc892da-dd53-43fd-9815-987aae968d7d">

abaevbog commented:
Right, sorry! Seems like I need to take a break now that I am confusing things

dstillman commented:
1. This doesn't technically do anything for accessibility on macOS until we do https://github.com/zotero/zotero/issues/4724 for the collection tree.
2. Instead of "Change Parent" and a super-awkward "Other Collections (As Copy)", how about Move To and Copy To? Move To would be limited to the current library, and Copy To could show libraries at the top level if there were any other than My Library.

(Copy To would also technically get us https://github.com/zotero/zotero/issues/1730. That might still be worth its own menu option, since navigating through the hierarchy could be annoying, but I don't know how common an operation duplicating a collection would be.)

abaevbog commented:
> Instead of "Change Parent" and a super-awkward "Other Collections (As Copy)", how about Move To and Copy To? Move To would be limited to the current library, and Copy To could show libraries at the top level if there were any other than My Library.

That sounds way better! I also agree that "Copy to" deserves its own independent menu option: I think it would look strange if "Copy to" was a submenu of "Move to".

How does this behavior look?

https://github.com/user-attachments/assets/850e400d-2875-4e2f-9b7e-fbaee05bb7b8

abaevbog commented:
Hm, I can't seem to reproduce it yet...

This is the setup, right?

https://github.com/user-attachments/assets/42d60575-6eef-4f38-a838-a29b8056d4d4

dstillman commented:
Yeah, that's the setup. It's listing the same key each time, and it's the collection with collectionID 1 in my DB. Something being cast incorrectly?

dstillman commented:
Or difference between libraryID 1 and collectionID 1?

abaevbog commented:
I believe this is what was wrong: `target.id` instead of something like `targetCollectionID: target instanceof Zotero.Collection ? target.id : null,` in the call to `ZoteroPane.collectionsView.executeCollectionCopy`.  I don't quite understand why it would not error on me because it does look like it should have... So I never could reproduce the issue, though the added test did fail without changing that line and it is passing now. Does it fix it for you?

dstillman commented:
That's working now, yes. I'd guess you weren't seeing it because you don't have a collection with collectionID 1.

abaevbog commented:
You are right, I just checked and I have no collection with id == 1.

abaevbog commented:
I rebased to clean up the tests merge conflicts

dstillman commented:
For "Move To", for a top-level collection, the current library should probably also be disabled, the same way that the parent collection is disabled for a subcollection.

dstillman commented:
This isn't checking destination editability (from `canDropCheck()`), nor does it look like it's checking things from `canDropCheckAsync()`.

We can just duplicate those things in ZoteroPane in order to get this merged, but later, we should probably refactor all of this copying code to get it out of the collection tree — maybe `Zotero.Collection::canCopy()`, `Zotero.Collection::canCopyAsync()`, and `Zotero.Collection::copy()`. (We already have `Zotero.Item::moveToLibrary()`, with fewer options.) And then for drag and drop, we'd just call those things from `canDropCheck()`/`canDropCheckAsync()` for `dataType == 'zotero/collection'`.

abaevbog commented:
> This isn't checking destination editability (from canDropCheck()), nor does it look like it's checking things from canDropCheckAsync().

Oh, right. Thank you for pointing it out. As far as I can see, there are three things that collectionTree checks for in `canDropCheck` and `canDropCheckAsync`:
1. Check if the group is editable: https://github.com/zotero/zotero/blob/c14896a64089ed3aed0d8bf04d1e9b3d24e4e3e4/chrome/content/zotero/collectionTree.jsx#L1569-L1572 I just added a small tweak for that, so targets that are non-editable groups will be disabled, so that should be taken care of.
3. Do not allow dropping a collection into itself or its children in canDropCheck: https://github.com/zotero/zotero/blob/c14896a64089ed3aed0d8bf04d1e9b3d24e4e3e4/chrome/content/zotero/collectionTree.jsx#L1687-L1697 We already disable those menuitems, so this should be good too.
4. Do not allow copying a collection into a library if a linked collection already exists: https://github.com/zotero/zotero/blob/c14896a64089ed3aed0d8bf04d1e9b3d24e4e3e4/chrome/content/zotero/collectionTree.jsx#L1776-L1793 That means that once you copy a collection into another library once, you cannot do it again. I just added an extra step to disable menus of groups containing subcollections. It happens after the "Copy to" menu is opened, since `ZoteroPane.buildCopyCollectionMenu` is not async, to avoid popup appearing after delay. Let me know if this makes sense!

I do have a few of questions about the last point:
- What is the role of linked collections? I see the relation is added but what is it used for? 
- I think it feels a bit off now that menus of libraries are disabled because I think it's not immediately clear why. One could copy a collection to a group, rename the copied collection, then come back some time later, and not be able to copy again, without a clear indication why. 
- I think "Copy to" now feels a tiny bit misleading, since now for all libraries outside of the current one it is "Copy (only once)". If it were "Create linked collection", or something like that, it might have clarified these concerns a bit but it doesn't work well for copying within the current library...
- Finally, if is important to maintain this one-to-one relations between linked collections, should we also add an extra check into `CollectionTree._copyCollections()` to see if a linked collection already exists and throw an error, if so? That is to avoid situations

dstillman commented on chrome/content/zotero/zoteroPane.js:
```diff
@@ -4091,13 +4091,38 @@ var ZoteroPane = new function()
+		// and disable their menuitems. Same logic as in CollectionTree.canDropCheckAsync.
+		let linkedCollectionsExist = {};
+		(async () => {
+			for (let library of Zotero.Libraries.getAll()) {
```
Start with `topLevelEntries` and remove the `instanceof Zotero.Feed` check below?

abaevbog replied:
Oh yeah, for sure! I had a glitch with the `async` execution, so I isolated `topLevelEntries` for troubleshooting. That had nothing to do with `topLevelEntries`, so I changed it back to `for (let library of topLevelEntries)`.

dstillman commented on chrome/content/zotero/zoteroPane.js:
```diff
@@ -4091,13 +4091,38 @@ var ZoteroPane = new function()
+				for (let descendent of selected.getDescendents(false, 'collection')) {
+					let subcollection = Zotero.Collections.get(descendent.id);
+					let linkedSubcollection = await subcollection.getLinkedCollection(library.libraryID, true);
+					linkedCollectionsExist[library.libraryID] = linkedSubcollection || linkedCollectionsExist[library.libraryID];
```
Clearer to just use `if (linkedSubcollection) {` instead of short-circuiting, I think

dstillman commented on chrome/content/zotero/zoteroPane.js:
```diff
@@ -4091,13 +4091,38 @@ var ZoteroPane = new function()
+			// Libraries that have linked collections have their menus disabled
+			for (let libraryMenuItem of [...popup.childNodes]) {
+				let menuItemLibID = libraryMenuItem.getAttribute("value").substring(1);
+				if (linkedCollectionsExist[parseInt(menuItemLibID)]) {
```
No real reason to use `parseInt` here as is, since the object key is always a string anyway. (Or you could use `Map` instead of an array for `linkedCollectionsExist`, and then they'd be actual numbers.)

dstillman commented on chrome/content/zotero/zoteroPane.js:
```diff
@@ -4135,7 +4160,8 @@ var ZoteroPane = new function()
 					return selected == target
-						|| (target instanceof Zotero.Group && !target.editable);
+						|| (target instanceof Zotero.Group && !target.editable)
+						|| linkedCollectionsExist[target.libraryID];
```
Is this supposed to be here? Is this not taken care of by the block above, which runs asynchronously?

abaevbog replied:
No, sorry about that. This was accidentally leftover from when I was trying to not disable menus asynchronously.

dstillman commented:
The link is currently used only to prevent a re-drag, but it'll be used in the future to support some sort of intelligent merging/overwriting of data between libraries, with some appropriate dialog (so these won't stay disabled). This is one of the biggest current limitations of group-based workflows, and while just blocking and forcing a deletion in the target library isn't a great solution, it lets us be prevent lots of duplicate copies for that eventual future.

dstillman commented:
> Finally, if is important to maintain this one-to-one relations between linked collections, should we also add an extra check into CollectionTree._copyCollections() to see if a linked collection already exists and throw an error, if so?

Not necessary. The front-end code shouldn't allow it in most cases, and I think you can end up with duplicates just by moving an item in the target library to the trash that and then undeleting it, so we'll need to handle the possibility of duplicate relations when we implement this properly in any case.

abaevbog commented:
> The link is currently used only to prevent a re-drag, but it'll be used in the future to support some sort of intelligent merging/overwriting of data between libraries, with some appropriate dialog (so these won't stay disabled). 

Gotcha! That makes sense, thank you for for the explanation.

While testing this out, I ran into the issue of finding it hard to connect which collection in another group is linked to the collection I am copying, especially if a subcollection is linked or if a collection got renamed after initial copying/linking. So, until the merging data logic is added, would we want to add some way to find out which collections are linked? As a quick fix, instead of disabling the group menu, we could set it to have a single disabled `menuitem` saying something like "Linked as {collection name}".  

<img width="521" alt="Screenshot 2024-10-14 at 3 15 43 PM" src="https://github.com/user-attachments/assets/dfffbf16-5cac-46e0-87ab-5a18b490f940">

At the same time, it doesn't seem like this drag-drop mechanism has caused much confusion in the past, so it very much might be redundant.
